cranelift/frontend/src/frontend.rs

```diff
@@ -610,11 +630,186 @@ impl<'a> FunctionBuilder<'a> {
+        //    live afterwards should need not be prevented from reclamation by
+        //    the GC for us, and therefore need not appear in this stack map. It
+        //    is the callee's responsibility to record such arguments in its
+        //    stack maps if it keeps them alive across some call that might
+        //    trigger GC.
+        //
+        // 3. Any use of a needs-stack-map value adds it to our `live` set.
+        //
+        //    Note: we do not flow liveness from block parameters back to branch
+        //    arguments, and instead always consider branch arguments live. That
+        //    additional precision would require a fixed-point loop.
+        //
+        //    Furthermore, we do not differentiate between uses of a
+        //    needs-stack-map value that ultimately flow into a side-effecting
+        //    operation versus uses that themselves are not live. This could be
+        //    tightened up in the future, but we're starting with the easiest,
+        //    simplest thing. Besides, none of our mid-end optimization passes
+        //    have run at this point in time yet, so there probably isn't much,
+        //    if any, dead code.
+        for block in self.func_ctx.dfs.post_order_iter(&self.func) {
+            let mut option_inst = self.func.layout.last_inst(block);
+            while let Some(inst) = option_inst {
+                // (1) Remove values defined by this instruction from the `live`
+                // set.
+                for val in self.func.dfg.inst_results(inst) {
+                    live.remove(val);
+                }
+
+                // (2) If this instruction is a call, then we need to add a
+                // safepoint to record any values in `live`.
+                let opcode = self.func.dfg.insts[inst].opcode();
+                if opcode.is_call() && !opcode.is_return() {
+                    let mut live: SmallVec<[_; 4]> = live.iter().copied().collect();
+                    live.sort_by_key(|val| self.func.dfg.value_type(*val).bytes());
+                    for chunk in live.chunk_by(|a, b| {
+                        self.func.dfg.value_type(*a).bytes() == self.func.dfg.value_type(*b).bytes()
+                    }) {
+                        let byte_size = self.func.dfg.value_type(chunk[0]).bytes();
+                        debug_assert!(byte_size.is_power_of_two());
+                        let index = usize::try_from(byte_size.ilog2()).unwrap();
+                        max_vals_in_stack_map_by_size[index] =
+                            core::cmp::max(max_vals_in_stack_map_by_size[index], chunk.len());
+                    }
+
+                    let old_val = safepoints.insert(inst, live);
+                    debug_assert!(old_val.is_none());
+                }
+
+                // (3) Add all needs-stack-map values that are operands to this
+                // instruction to the live set.
```

Comment:
Can we add a comment here noting that this (implicitly) includes block args as well, in reference to above discussion of blockparams? Was looking for the logic as an explicit thing for branches and missed this (implicit/more general) handling.